lib/service/certreloader.go

```diff
@@ -0,0 +1,158 @@
+type CertReloaderConfig struct {
+	// KeyPairs are the key and certificate pairs that the proxy will load.
+	KeyPairs []KeyPairPath
+	// KeyPairsReloadInterval is the interval between attempts to reload
+	// x509 key pairs. If set to 0, then periodic reloading is disabled.
+	KeyPairsReloadInterval time.Duration
+}
+
+// CertReloader periodically reloads a list of cert key-pair paths.
+// This allows new certificates to be used without a full reload of Teleport.
+type CertReloader struct {
+	*log.Entry
+	// cfg is the certificate reloader configuration.
+	cfg CertReloaderConfig
+
+	// certificates is the list of certificates loaded.
+	certificates []tls.Certificate
+	// mu protects the list of certificates.
+	mu sync.RWMutex
+}
+
+// NewCertReloader initializes a new certificate reloader.
+func NewCertReloader(cfg CertReloaderConfig) *CertReloader {
+	return &CertReloader{
+		Entry: log.WithFields(log.Fields{
+			trace.Component: teleport.Component(teleport.ComponentProxy, "certreloader"),
+		}),
+		cfg: cfg,
+	}
+}
+
+// Run tries to load certificates and then spawns the certificate reloader.
+func (c *CertReloader) Run(ctx context.Context) error {
+	// Synchronously load initially configured certificates.
+	if err := c.loadCertificates(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Do not reload certificates if the interval is 0 (the default).
+	if c.cfg.KeyPairsReloadInterval == 0 {
+		return nil
+	}
+
+	// Spawn the certificate reloader.
+	go func() {
+		c.Infof("Starting periodic reloading of certificate key pairs every %s.", c.cfg.KeyPairsReloadInterval)
+		defer func() {
+			c.Info("Stopped periodic reloading of certificate key pairs.")
+		}()
+
+		t := time.NewTicker(c.cfg.KeyPairsReloadInterval)
+		defer t.Stop()
+		for {
+			select {
+			case <-t.C:
+				if err := c.loadCertificates(); err != nil {
+					c.WithError(err).Warn("Failed to load certificates")
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+	return nil
+}
+
+// loadCertificates loads certificate keys pairs.
+// It returns an error if any of the certificate key pairs fails to load.
+// If any of the key pairs fails to load, none of the certificates are updated.
+func (c *CertReloader) loadCertificates() error {
+	certs := make([]tls.Certificate, 0, len(c.cfg.KeyPairs))
+	for _, pair := range c.cfg.KeyPairs {
+		c.Debugf("Loading TLS certificate %v and key %v.", pair.Certificate, pair.PrivateKey)
+
+		certificate, err := tls.LoadX509KeyPair(pair.Certificate, pair.PrivateKey)
+		if err != nil {
+			// If one certificate fails to load, then no certificate is updated.
+			return trace.Wrap(err)
+		}
+
+		// Parse the end entity cert and add it to certificate.Leaf.
+		// With this, the SupportsCertificate call doesn't have to
+		// parse it on every GetCertificate call.
+		leaf, err := x509.ParseCertificate(certificate.Certificate[0])
+		if err != nil {
+			// If one certificate fails to load, then no certificate is updated.
+			return trace.Wrap(err)
```

Comment:
Maybe these wraps ought to include a message or field that indicates which certificate path has failed, otherwise, users with multiple configured certificates will have to work it out themselves which is causing the issue.

Reply:
Thanks for the review! Done in b79840d120a6975ac21b77c35e349c3429a63bd6.